toolchain/diagnostics/diagnostic_converter.h

```diff
@@ -27,9 +27,16 @@ enum class DiagnosticTypeConversion : int8_t {
 template <typename LocationT>
 class DiagnosticConverter {
  public:
+  using ContextFnT = llvm::function_ref<void(
+      DiagnosticLocation, const Internal::DiagnosticBase<>&)>;
+
   virtual ~DiagnosticConverter() = default;
 
-  virtual auto ConvertLocation(LocationT loc) const -> DiagnosticLocation = 0;
+  // Converts a LocationT to a DiagnosticLocation. The function may optionally
+  // provide context messages using context_fn, but these must be simple and
+  // already have converted locations.
```

Comment:
I'm having trouble following this comment, but I can't tell if that's because I'm not familiar with this code or if it needs clarification. What are these context messages, and what do they have to do with the conversion? What counts as "simple"? Isn't the fact that they already have converted locations guaranteed by the fact that `context_fn` takes a `DiagnosticLocation` as its argument?

Also, this comment seems to be expressed in terms of requirements on the implementation ("...these _must_ be simple...") and permission given to the implementation ("the function may _optionally_..."). I think it would be more conventional to express it in terms of requirements on/permissions to the caller, and guarantees made by the implementation, e.g. "ConvertLocation may invoke context_fn to provide context messages, but they will be simple, and already have converted locations."

Reply:
Done, dropping "simple".

Reply:
Is it possible for a `DiagnosticLocation` to not be a "converted location"?

Reply:
If I understand what you're trying to ask, while a `DiagnosticLocation` could in theory be tracked directly, it's unlikely and not how the code works.

Reply:
I'm trying to understand what "they must already have converted locations" means. In this context, I would ordinarily assume it means that their locations must already have been converted to `DiagnosticLocation`, but the relevant parameter of `context_fn` has type `DiagnosticLocation`, so that statement seems equivalent to "calls to `context_fn` must type-check". So I feel like I must have misunderstood something; my question was a guess at where I went wrong, but I don't understand your answer.

(By the way, in line with my original comment, I suggest changing "must" to "will").

Reply:
> I'm trying to understand what "they must already have converted locations" means. In this context, I would ordinarily assume it means that their locations must already have been converted to `DiagnosticLocation`, but the relevant parameter of `context_fn` has type `DiagnosticLocation`, so that statement seems equivalent to "calls to `context_fn` must type-check". So I feel like I must have misunderstood something; my question was a guess at where I went wrong, but I don't understand your answer.

That's essentially what it's saying. I think the difference is that I'm expecting this to be more subtle than you; I expect callers might be surprised given that every other API deals with LocationT, and I'm emphasizing that this is a deliberate choice through the comment.

> (By the way, in line with my original comment, I suggest changing "must" to "will").

Maybe you could clarify why? I think "must" as a good grammatical match for "may": when something _may_ happen, then it _must_ occur in a given way; it's an obligation. For contrast, "will" implies inevitability that doesn't exist due to "may".

Reply:
> I expect callers might be surprised given that every other API deals with LocationT, and I'm emphasizing that this is a deliberate choice through the comment.

In that case, it might be clearer as a comment on `ContextFnT`, where it can talk about the parameter type more directly. I've left a suggestion for doing that.

> > (By the way, in line with my original comment, I suggest changing "must" to "will").
>
> Maybe you could clarify why? I think "must" as a good grammatical match for "may": when something _may_ happen, then it _must_ occur in a given way; it's an obligation. For contrast, "will" implies inevitability that doesn't exist due to "may". [when...will]
> 
> Note, some of this may come down to personal writing styles.

This is a moot point if you accept my suggestion above, but in case it's helpful: in my experience, people typically use language of factual assertion to give guarantees and take on obligations (e.g. "The project _will_ be done by Friday"), and use language of obligation and permission to impose requirements on others (e.g. "The project _must_ be done by Friday"). I automatically read function comments as speaking to callers on behalf of the implementers, not the other way around, so when a comment says "must", I assume that it's imposing a requirement on the callers, not making a guarantee to them (i.e. imposing a requirement on implementers). Deviating from that convention is particularly rough for me in this case, because the inversion of control makes it harder than usual to work out the intent from context.

My suggestion of using "may" caused confusion here because it can be read as granting permission, or as stating a factual possibility (like "I may work from home tomorrow"). The latter is what I intended, but it might have been better to use "can" or "might" instead, to reduce the ambiguity.

Reply:
Adopted the other suggested comment on ContextFn, removed this clause.